src/core/qgsvectorlayerlabeling.cpp

```diff
@@ -88,3 +92,454 @@ QgsVectorLayerSimpleLabeling *QgsVectorLayerSimpleLabeling::create( const QDomEl
 
   return new QgsVectorLayerSimpleLabeling( QgsPalLayerSettings() );
 }
+
+QPointF quadOffsetToSLDAnchor( QgsPalLayerSettings::QuadrantPosition quadrantPosition )
+{
+  double quadOffsetX = 0.5, quadOffsetY = 0.5;
+
+  // adjust quadrant offset of labels
+  switch ( quadrantPosition )
+  {
+    case QgsPalLayerSettings::QuadrantAboveLeft:
+      quadOffsetX = 1;
+      quadOffsetY = 0;
+      break;
+    case QgsPalLayerSettings::QuadrantAbove:
+      quadOffsetX = 0.5;
+      quadOffsetY = 0;
+      break;
+    case QgsPalLayerSettings::QuadrantAboveRight:
+      quadOffsetX = 0;
+      quadOffsetY = 0;
+      break;
+    case QgsPalLayerSettings::QuadrantLeft:
+      quadOffsetX = 1;
+      quadOffsetY = 0.5;
+      break;
+    case QgsPalLayerSettings::QuadrantRight:
+      quadOffsetX = 0;
+      quadOffsetY = 0.5;
+      break;
+    case QgsPalLayerSettings::QuadrantBelowLeft:
+      quadOffsetX = 1;
+      quadOffsetY = 1;
+      break;
+    case QgsPalLayerSettings::QuadrantBelow:
+      quadOffsetX = 0.5;
+      quadOffsetY = 1;
+      break;
+    case QgsPalLayerSettings::QuadrantBelowRight:
+      quadOffsetX = 0;
+      quadOffsetY = 1.0;
+      break;
+    case QgsPalLayerSettings::QuadrantOver:
+    default:
+      break;
+  }
+
+  return QPointF( quadOffsetX, quadOffsetY );
+}
+
+QgsUnitTypes::RenderUnit toRenderUnits( QgsPalLayerSettings::SizeUnit palUnit )
+{
+  switch ( palUnit )
+  {
+    case QgsPalLayerSettings::MM:
+      return QgsUnitTypes::RenderMillimeters;
+    case QgsPalLayerSettings::Points:
+      return QgsUnitTypes::RenderPoints;
+    case QgsPalLayerSettings::MapUnits:
+      return QgsUnitTypes::RenderMapUnits;
+    default:
+      return QgsUnitTypes::RenderPixels;
+  }
+}
+
+/*
+ * This is not a generic function encoder, just enough to encode the label case control functions
+ */
+void appendSimpleFunction( QDomDocument doc, QDomElement parent, QString name, QString attribute )
+{
+  QDomElement function = doc.createElement( QStringLiteral( "ogc:Function" ) );
+  function.setAttribute( QStringLiteral( "name" ), name );
+  parent.appendChild( function );
+  QDomElement property = doc.createElement( QStringLiteral( "ogc:PropertyName" ) );
+  property.appendChild( doc.createTextNode( attribute ) );
+  function.appendChild( property );
+}
+
+std::unique_ptr<QgsMarkerSymbolLayer> backgroundToMarkerLayer( QgsTextBackgroundSettings settings )
+{
+  std::unique_ptr<QgsMarkerSymbolLayer> layer;
+  if ( settings.type() == QgsTextBackgroundSettings::ShapeSVG )
```

Comment:
Can you make this an explicit switch (with no default)? It's a bit more future proof in case any addition shapes are added in future